pgvectorscale/src/access_method/sbq.rs

```diff
@@ -527,43 +544,70 @@ impl<'a> SbqSpeedupStorage<'a> {
                     }
 
                     let rn_neighbor = unsafe {
-                        SbqNode::read(self.index, neighbor_index_pointer, &mut lsr.stats)
+                        SbqNode::read(
+                            self.index,
+                            neighbor_index_pointer,
+                            self.has_labels,
+                            &mut lsr.stats,
+                        )
                     };
+
                     let node_neighbor = rn_neighbor.get_archived_node();
                     let bq_vector = node_neighbor.get_bq_vector();
-                    let distance = lsr.sdm.as_ref().unwrap().calculate_bq_distance(
-                        bq_vector,
-                        gns,
-                        &mut lsr.stats,
-                    );
+                    let distance = lsr
+                        .sdm
+                        .as_ref()
+                        .expect("sdm is Some")
+                        .calculate_bq_distance(bq_vector, gns, &mut lsr.stats);
+
+                    // Skip neighbors that have no matching labels with the query
+                    if let Some(labels) = lsr.sdm.as_ref().expect("sdm is Some").query.labels() {
+                        if !labels
+                            .overlaps(node_neighbor.get_labels().expect("Unlabeled neighbor?"))
+                        {
+                            continue;
+                        }
+                    }
 
                     let lsn = ListSearchNeighbor::new(
                         neighbor_index_pointer,
                         lsr.create_distance_with_tie_break(distance, neighbor_index_pointer),
                         PhantomData::<bool>,
+                        node_neighbor.get_labels().map(Into::into),
                     );
 
                     lsr.insert_neighbor(lsn);
                 }
             }
             GraphNeighborStore::Builder(b) => {
-                let neighbors = b.get_neighbors(lsn_index_pointer);
-                for &neighbor_index_pointer in neighbors.iter() {
+                let mut neighbors: Vec<NeighborWithDistance> = Vec::new();
+                b.get_neighbors_with_full_vector_distances(lsn_index_pointer, &mut neighbors);
+                for neighbor in neighbors.iter() {
+                    let neighbor_index_pointer = neighbor.get_index_pointer_to_neighbor();
                     if !lsr.prepare_insert(neighbor_index_pointer) {
                         continue;
                     }
                     let mut cache = self.qv_cache.borrow_mut();
-                    let bq_vector = cache.get(neighbor_index_pointer, self, &mut lsr.stats);
-                    let distance = lsr.sdm.as_ref().unwrap().calculate_bq_distance(
-                        bq_vector,
-                        gns,
-                        &mut lsr.stats,
-                    );
+                    let (bq_vector, _) = cache.get(neighbor_index_pointer, self, &mut lsr.stats);
```

Comment:
shouldn't we be using the labels from the cache here instead of `neighbor.get_labels()` below?